File: built-in plugins dsp/mb-time-scatter.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbTimeScatterPlugin: PluginDefinition = { id: 'mb-time-scatter', slug: 'mb-time-scatter', name: 'MB Scatter', category: 'effect', type: 'delay' as any, version: '1.0.0', description: 'Tempo-synced audio scatter with randomized playback order', author: 'Max Booster', parameters: [{ id: 'division', name: 'Division', type: 'float', defaultValue: 8, minValue: 2, maxValue: 32, automatable: false }, { id: 'probability', name: 'Probability', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'reverse', name: 'Reverse', type: 'float', defaultValue: 0.2, minValue: 0, maxValue: 1, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { division: 8, probability: 0.3, reverse: 0.2, mix: 0.5 } };

export default MbTimeScatterPlugin;

File: built-in plugins dsp/mb-time-scatter.cpp
/**
 * MB Scatter
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Tempo-synced audio scatter with randomized playback order
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_TIME_SCATTER_H
#define MB_TIME_SCATTER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbTimeScatter : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-time-scatter";
    static constexpr const char* PLUGIN_NAME    = "MB Scatter";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float division = 8f;  // range [2, 32]
    float probability = 0.3f;  // range [0, 1]
    float reverse = 0.2f;  // range [0, 1]
    float mix = 0.5f;  // range [0, 1]
    };

    MbTimeScatter() = default;
    ~MbTimeScatter() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.division = std::clamp(params.division, 2f, 32f);
        params.probability = std::clamp(params.probability, 0f, 1f);
        params.reverse = std::clamp(params.reverse, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Scatter
        return input;
    }
};

#endif // MB_TIME_SCATTER_H
